### PR TITLE
Reject substituting extension schemas or owners matching ["$'\].

### PR DESCRIPTION
**Pull request by divyeshddv**
[Reject substituting extension schemas or owners matching ["$'\].](https://github.com/greenplum-db/gpdb/pull/17302/commits/36b9b0e10184635ec0b6e2a15cdf7d643e7d1ade) 

Substituting such values in extension scripts facilitated SQL injection
when @extowner@, @extschema@ appeared inside a
quoting construct (dollar quoting, '', or "").  No bundled extension was
vulnerable.  Vulnerable uses do appear in a documentation example and in
non-bundled extensions.  Hence, the attack prerequisite was an
administrator having installed files of a vulnerable, trusted,
non-bundled extension.  Subject to that prerequisite, this enabled an
attacker having database-level CREATE privilege to execute arbitrary
code as the bootstrap superuser.  By blocking this attack in the core
server, there's no need to modify individual extensions.  Back-patch to
v11 (all supported versions).

Reported by Micah Gate, Valerie Woolard, Tim Carey-Smith, and Christoph
Berg.
Security: CVE-2023-39417
(cherry picked from commit cd5f2a357014d5ed8205191f8e1fb180f1439599)

cherry-pick diffs: exclude diffs due to @extschema:...@ syntax
which we don't support on 6x

### CODE REVIEW

File: src/backend/commands/extension.c [hunk 1]
```diff
@@ -935,12 +945,19 @@ execute_extension_script(Node *stmt,
...
 
-			t_sql = DirectFunctionCall3(replace_text,
-										t_sql,
-										CStringGetTextDatum("@extschema@"),
-										CStringGetTextDatum(qSchemaName));
+			t_sql = DirectFunctionCall3Coll(replace_text,
+											C_COLLATION_OID,
+											t_sql,
+											CStringGetTextDatum("@extschema@"),
+											CStringGetTextDatum(qSchemaName));
```
Comment (soumyadeep2007) [inline src/backend/commands/extension.c:955]: In 6X, we don't have changes from https://github.com/greenplum-db/gpdb/commit/5e1963fb764e9cc092e0f7b58b28985c311431d9. So it is not strictly necessary to have this diff I feel.

You can still have the DirectFunctionCall3 and have the if condition right?

**Review by soumyadeep2007 — COMMENTED** [review]
Couple things:

1.Lets add it to the commit message that we have skipped https://github.com/postgres/postgres/commit/cd5f2a357014d5ed8205191f8e1fb180f1439599#diff-f8bf486566de0369065b520e001b0434577000d53dacd135f733931adabc0883R1070-R1093 as the syntax doesn't apply.
2. I couldn't find the commit linked in the commit message:
`(cherry picked from commit eb044d8f0aee1ba4950b0867f6ca9328374318db)`
Should we use cd5f2a357014d5ed8205191f8e1fb180f1439599 instead?

File: src/backend/commands/extension.c (hunk 1 above)
Comment (divyeshddv) [reply src/backend/commands/extension.c]: Yes, we can do that, removed the diff

**Comment by divyeshddv** [comment]
> 1.Lets add it to the commit message that we have skipped [postgres/postgres@cd5f2a3#diff-f8bf486566de0369065b520e001b0434577000d53dacd135f733931adabc0883R1070-R1093](https://github.com/postgres/postgres/commit/cd5f2a357014d5ed8205191f8e1fb180f1439599#diff-f8bf486566de0369065b520e001b0434577000d53dacd135f733931adabc0883R1070-R1093) as the syntax doesn't apply. 

Done!

>2. I couldn't find the commit linked in the commit message: `(cherry picked from commit eb044d8f0aee1ba4950b0867f6ca9328374318db)` Should we use cd5f2a357014d5ed8205191f8e1fb180f1439599 instead?

Changed, I think it came from the 7x version on my branch that I cherrypicked back

**Review by soumyadeep2007 — APPROVED** [review]
LGTM!

**Pull request merged**